src/java/org/apache/cassandra/service/disk/usage/DiskUsageMonitor.java

```diff
@@ -0,0 +1,229 @@
+        this.dataDirectoriesSupplier = DiskUsageMonitor::dataDirectoriesGroupedByFileStore;
+    }
+
+    @VisibleForTesting
+    public DiskUsageMonitor(Supplier<Multimap<FileStore, Directories.DataDirectory>> dataDirectoriesSupplier)
+    {
+        this.dataDirectoriesSupplier = dataDirectoriesSupplier;
+    }
+
+    /**
+     * Start monitoring local disk usage and call notifier when local disk usage state changed.
+     */
+    public void start(Consumer<DiskUsageState> notifier)
+    {
+        // start the scheduler regardless guardrail is enabled, so we can enable it later without a restart
+        ScheduledExecutors.scheduledTasks.scheduleAtFixedRate(() -> {
+
+            if (!Guardrails.localDataDiskUsage.enabled(null))
+                return;
+
+            updateLocalState(getDiskUsage(), notifier);
+        }, 0, CassandraRelevantProperties.DISK_USAGE_MONITOR_INTERVAL_MS.getLong(), TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    public void updateLocalState(double usageRatio, Consumer<DiskUsageState> notifier)
+    {
+        double percentage = usageRatio * 100;
+        long percentageCeiling = (long) Math.ceil(percentage);
+
+        DiskUsageState state = getState(percentageCeiling);
+
+        Guardrails.localDataDiskUsage.guard(percentageCeiling, state.toString(), false, null);
+
+        // if state remains unchanged, no need to notify peers
+        if (state == localState)
+            return;
+
+        localState = state;
+        notifier.accept(state);
+    }
+
+    /**
+     * @return local node disk usage state
+     */
+    @VisibleForTesting
+    public DiskUsageState state()
+    {
+        return localState;
+    }
+
+    /**
+     * @return disk usage (including all memtable sizes) ratio
+     */
+    @VisibleForTesting
+    public double getDiskUsage()
+    {
+        // using BigDecimal to handle large file system
+        BigDecimal used = BigDecimal.ZERO; // space used by data directories
+        BigDecimal usable = BigDecimal.ZERO; // free space on disks
+
+        for (Map.Entry<FileStore, Collection<Directories.DataDirectory>> e : dataDirectoriesSupplier.get().asMap().entrySet())
+        {
+            usable = usable.add(BigDecimal.valueOf(usableSpace(e.getKey())));
+
+            for (Directories.DataDirectory dir : e.getValue())
+                used = used.add(BigDecimal.valueOf(dir.getRawSize()));
+        }
+
+        // The total disk size for data directories is the space that is actually used by those directories plus the
+        // free space on disk that might be used for storing those directories in the future.
+        BigDecimal total = used.add(usable);
+
+        // That total space can be limited by the config property data_disk_usage_max_disk_size.
+        DataStorageSpec diskUsageMaxSize = guardrailsConfigSupplier.get().getDataDiskUsageMaxDiskSize();
+        if (diskUsageMaxSize != null)
+            total = total.min(BigDecimal.valueOf(diskUsageMaxSize.toBytes()));
```

Comment:
One more thing. Maybe a good thing will be to have a getter for the value in the default internal unit; returns the parameter value in the unit we should use internally which is the one used in Config class. In this case we use Bytes. Then everywhere internally we can use that getter ensuring we don't get tempted to use wrong conversion, going up for example and loosing precision. The reason why we have the smallest unit introduced for users to ensure they don't run into that problem with the old parameters I transferred without changing what the parameters use internally. Every parameter is different of course so we always need to think what is the best for the particular one. Nothing has changed from before in regards to thinking what should be the boundaries for a parameter that we want to accept and what units we want to encourage or not, except time/storage and rate are not negative things.. At the end of the day the new types are only wrappers that we need for a  parser but upper boundary is something we still need to consider on a per case basis. 
I am actually thinking maybe to move to abstract Boundary classes and split into Int and long ones to make people think more in particular about that... That is a topic for another ticket of course, just sharing thoughts.

Reply:
In this particular case the conversion, which is a series of relatively simple arithmetic operations, happens once when the periodic monitoring task is run. The default rate is 30 seconds, so having a `toBytes` every 30 seconds doesn't seem very problematic compared to checks on the hot path. Probably the `toBytes` call done in the guardrail for collection size could be more problematic, since it's done for every collection cell during writes.

I guess that memoizing the number of bytes into `DataStorageSpec` (or a subclass of it) could improve possible performance problems. If I'm right `DataStorageSpec` is only used in config so we are not going to have many instances on memory, so memoizing is not going to create memory issues. Also, we would only need this optimization on size properties on the hot path.

Performance-wise, an alternative approach would be splitting `Config` in two classes, one for parsing and other for holding the status. The parsing class, say `YamlConfig`, would be short-lived and only used to parse the yaml at startup. It would contain the `DataStorageSpec` instances and other parse-only classes. The class holding the status, say `ConfigStatus`, would be a singleton containing the `volatile` properties in the format that they are going to be internally consumed, such as a longs for bytes. JMX methods would directly write into that `ConfigStatus` class. That's just idea, and I'm not sure it would worth the additional complexity.

Reply:
I had some conversation with @blambov yesterday and he doesn't think that single multiplication (we don't do division to prevent any precision problems, that's why we set minimum unit) especially when we use 1024 would have any impact. 
Regarding `DataStorageSpec` - yes, it is used only by config. @blambov any concerns in regard to the exact guardrail mentioned - collection, since it's done for every collection cell during writes.

Reply:
In this particular piece of code, I would consider the `Config` conversions trivial compared to the cost of operating on `BigDecimal`. Any of the additions in the code above is at least one, possibly two orders of magnitude more complex than that conversion.

I would not worry about the configuration changes. Outside the scope of this ticket, it may be a good idea to consider why a `BigDecimal` is required instead of `BigInteger` or `long`, possibly pre-multiplied by some fixed power of two.

Reply:
Thanks for the feedback. Yep, the conversion config properties doesn't look like a big problem here, since it happens in the middle of `BigDecimal` operations and the calculation of the directories size. Besides, this is a periodic task which triggers every 30s, different to other config accesses in the hot path.

As for the use of `BigDecimal`, I guess we can use `BigInteger` until the final division. We use those types instead of double/long to avoid problems while managing very large file systems.

Reply:
Thanks @blambov , actually the question was more about the collection guardrail that was mentioned. l guess we can discuss this separately from this ticket but as it was slightly touched here in this patch (prohibit 0) and there is no ticket (as there is uncertainty whether there is a real issue to handle or not) I was just interested to get an opinion on that one.